### PR TITLE
Add support for more numeric types to oximeter

### PR DESCRIPTION
**Pull request by bnaecker**
- Adds support for all integral widths to fields
- Adds support for all integral widths to gauge and cumulative types
- Adds support for all integral widths to the histogram types
- Reworks methods for generating log-linear bins. This supports generating histograms with bin counts other than 10 and using bases other than 10. The handling of floating point values has also been reworked, including several stringent tests against a reference implementation for linearly-spaced values (NumPy). This is all pretty tricky given the well-known difficulties of comparing floating point types for equality, and we may want to avoid using floats for histogram supports in the future. Cautionary warnings in the type's documentation have been added in the meantime.
- Adds support for new types to the ClickHouse client, including tests for serialization of fields and measurements to the actual tables.
- Adds some USDT probes at the start / end of queries, including raw SQL.

### CODE REVIEW

**Comment by bnaecker** [comment]
Resolves #4058. Although this PR is pretty huge, it's also mostly mechanical. I've added support for: fields and measurements for all the integer types; cumulative `u64` and `f64` measurements; histograms for all integer types; and histograms for `f32`s. The meat of the PR is in the histogram code, to support bases and bins-per-magnitude other than an even 10.

This also kind of brings up an interesting question: should we support floating-point types at all? They're obviously tricky, but also very useful. Without them, a unit / resolution is basically required for anything meaningful, and those are not yet in place. There are a few interacting problems with them: difficulty of comparison; rounding errors; and ClickHouse's pretty lax formatting rules, which seem to print at most 7-8 digits of precision. I've opted to keep them so far, but I'm happy to revisit that.

File: oximeter/oximeter/src/traits.rs [hunk 1]
```diff
@@ -199,85 +207,87 @@ pub trait Datum: Clone {
...
 impl Cumulative for f64 {}
 
 /// A trait identifying types used as gauges
 pub trait Gauge: Datum {}
 
-impl Gauge for String {}
-impl Gauge for bool {}
-impl Gauge for i64 {}
-impl Gauge for f64 {}
+macro_rules! impl_guage {
```
Comment (smklein) [inline oximeter/oximeter/src/traits.rs:283]: Nit: Spelling (guage vs gauge)

**Review by smklein — APPROVED** [review]

File: oximeter/db/src/client.rs [hunk 2]
```diff
@@ -3,24 +3,45 @@
...
+use std::collections::BTreeSet;
 use std::convert::TryFrom;
 use std::net::SocketAddr;
 use std::num::NonZeroU32;
 use std::sync::Mutex;
 use uuid::Uuid;
 
+#[usdt::provider(provider = "clickhouse__client")]
+mod probes {
+    fn query__start(_: &usdt::UniqueId, sql: &str) {}
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply oximeter/db/src/client.rs:41]: Good question! That is used to turn the probe name into `kebab-case`. This is a mechanism that's been in DTrace itself for a long time, where it translates dunders into dashes. It's really just aesthetic, but common in DTrace names (e.g., `proc:::exec-success`).

File: oximeter/oximeter/src/traits.rs (hunk 1 above)
Comment (bnaecker) [reply oximeter/oximeter/src/traits.rs:283]: Good catch thanks. Will fix shortly.

File: oximeter/db/src/client.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmustacc) [reply oximeter/db/src/client.rs:41]: The reason this feature exists in D is that many languages don't support the `-` character in identifier names. This is true for C. I think it's true for Rust (based on doc reading). And yeah as Ben points the use of hyphens is stylistically what is used to distinguish things in probe point names.

**Pull request merged**